test/integration/matrix/download.go

```diff
@@ -0,0 +1,122 @@
+package test
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// latestReleases for 'product' it checks throught the last 20 releases for the
+// 3 most recent minor versions (semantic versioning, major.MINOR.micro).
+func latestReleases(product string) []string {
+	jsonBody := lastTwenty(product, "")
+
+	type jsonVersion struct {
+		Version string
+	}
+	var versions []jsonVersion
+	if err := json.Unmarshal(jsonBody, &versions); err != nil {
+		panic(err)
+	}
+
+	unique := make(map[string]struct{}, 3)
+	result := []string{}
+	for _, v := range versions {
+		if len(unique) >= 3 {
+			break
+		}
+		vl := strings.Split(v.Version, ".")
+		minorVersion := vl[1]
+		if _, ok := unique[minorVersion]; !ok {
+			result = append(result, v.Version)
+			unique[minorVersion] = struct{}{}
+		}
+	}
+
+	return result
+}
+
+// getBinary returns the path to the binary for the requested product/version
+// if the binary is not installed locally, it is downloaded.
+func getBinary(product, version string) string {
+	const binPath = "./testdata/bin/%s/%s"
+	url := downloadURL(product, version)
+	path := fmt.Sprintf(binPath, product, version)
+	os.MkdirAll(filepath.Dir(path), 0755)
+	downloadBinary(url, path)
+	return path
+}
+
+// downloadURL returns the URL (string) to the (zipped) binary for AMD64
+// for a given product and version (eg. vault and 1.13.2).
+func downloadURL(product, version string) string {
+	jsonBody := aboutRelease(product, version)
+	type build struct {
+		Arch, OS, URL string
+	}
+	type releasesJSON struct {
+		Builds []build
+	}
+	var rel releasesJSON
+	if err := json.Unmarshal(jsonBody, &rel); err != nil {
+		panic(err)
+	}
+
+	for _, b := range rel.Builds {
+		if b.Arch == "amd64" && b.OS == "linux" {
```

Comment:
Do you think making this work for other platforms would be as simple as swapping this with something like this?

```
		if b.Arch == runtime.GOARCH && b.OS == runtime.GOOS {
```

(and similar in the runner code)

Reply:
Sure.. something like that should work. Though multi-platform testing is not in scope at this time.